File: MoneyMind_FlutterApp/lib/services/community_services.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';

import '../models/community_models.dart';

class CommunityService {
  final String baseUrl;

  CommunityService({required this.baseUrl});

  // User endpoints
  Future<User> createUser(String username, String handle, {String? bio}) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/community/users'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'username': username,
          'handle': handle,
          'bio': bio,
        }),
      ).timeout(const Duration(seconds: 60)); // Increased timeout

      if (response.statusCode == 201) {
        return User.fromJson(jsonDecode(response.body));
      } else {
        throw Exception('Failed to create user: ${response.body}');
      }
    } catch (e) {
      print('Error in createUser: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<UserDetail> getUserDetails(int userId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/users/$userId'),
      ).timeout(const Duration(seconds: 60)); // Increased timeout

      if (response.statusCode == 200) {
        return UserDetail.fromJson(jsonDecode(response.body));
      } else {
        throw Exception('Failed to get user details: ${response.body}');
      }
    } catch (e) {
      print('Error in getUserDetails: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<User> updateUser(int userId, {String? username, String? bio}) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl/api/community/users/$userId'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          if (username != null) 'username': username,
          if (bio != null) 'bio': bio,
        }),
      ).timeout(const Duration(seconds: 60)); // Increased timeout

      if (response.statusCode == 200) {
        return User.fromJson(jsonDecode(response.body));
      } else {
        throw Exception('Failed to update user: ${response.body}');
      }
    } catch (e) {
      print('Error in updateUser: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<String> uploadProfileImage(int userId, File imageFile) async {
    try {
      var request = http.MultipartRequest(
        'POST',
        Uri.parse('$baseUrl/api/community/users/$userId/profile-image'),
      );

      final fileName = imageFile.path.split('/').last;
      final fileExtension = fileName.split('.').last.toLowerCase();

      request.files.add(
        await http.MultipartFile.fromPath(
          'file',
          imageFile.path,
          contentType: MediaType('image', fileExtension),
        ),
      );

      final streamedResponse = await request.send().timeout(const Duration(seconds: 120)); // Extended timeout for uploads
      final response = await http.Response.fromStream(streamedResponse);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final filePath = data['file_path'];

        // Convert relative path to absolute URL if needed
        if (filePath.startsWith('/')) {
          return '$baseUrl$filePath';
        }
        return filePath;
      } else {
        throw Exception('Failed to upload profile image: ${response.body}');
      }
    } catch (e) {
      print('Error in uploadProfileImage: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<void> followUser(int userId) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/community/users/$userId/follow'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode != 204) {
        throw Exception('Failed to follow user: ${response.body}');
      }
    } catch (e) {
      print('Error in followUser: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<void> unfollowUser(int userId) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/api/community/users/$userId/follow'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode != 204) {
        throw Exception('Failed to unfollow user: ${response.body}');
      }
    } catch (e) {
      print('Error in unfollowUser: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<List<User>> getUserFollowers(int userId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/users/$userId/followers'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        return data.map((user) => User.fromJson(user)).toList();
      } else {
        throw Exception('Failed to get followers: ${response.body}');
      }
    } catch (e) {
      print('Error in getUserFollowers: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<List<User>> getUserFollowing(int userId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/users/$userId/following'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        return data.map((user) => User.fromJson(user)).toList();
      } else {
        throw Exception('Failed to get following: ${response.body}');
      }
    } catch (e) {
      print('Error in getUserFollowing: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<List<User>> searchUsers(String query) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/users/search?query=$query'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        return data.map((user) => User.fromJson(user)).toList();
      } else {
        throw Exception('Failed to search users: ${response.body}');
      }
    } catch (e) {
      print('Error in searchUsers: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<Post> createPost(String content, {File? imageFile}) async {
    try {
      print('Creating post with content: $content');
      print('Image file: ${imageFile?.path}');

      var request = http.MultipartRequest(
        'POST',
        Uri.parse('$baseUrl/api/community/posts'),
      );

      request.fields['content'] = content;

      if (imageFile != null) {
        print('Attaching image file');
        final fileName = imageFile.path.split('/').last;
        final fileExtension = fileName.split('.').last.toLowerCase();

        request.files.add(
          await http.MultipartFile.fromPath(
            'image',
            imageFile.path,
            contentType: MediaType('image', fileExtension),
          ),
        );
      }

      print('Sending request to: ${request.url}');
      final streamedResponse = await request.send().timeout(const Duration(seconds: 120)); // Extended timeout for uploads
      final response = await http.Response.fromStream(streamedResponse);

      print('Response status: ${response.statusCode}');
      print('Response body: ${response.body}');

      if (response.statusCode == 201) {
        final postData = jsonDecode(response.body);

        // Fix image URL if it's a relative path
        if (postData['image_url'] != null && postData['image_url'].toString().startsWith('/')) {
          postData['image_url'] = '$baseUrl${postData['image_url']}';
        }

        final post = Post.fromJson(postData);
        print('Post created successfully with ID: ${post.id}');
        return post;
      } else {
        print('Error response: ${response.statusCode} - ${response.body}');
        throw Exception('Failed to create post: ${response.body}');
      }
    } catch (e) {
      print('Exception in createPost: $e');
      throw Exception('Error creating post: $e');
    }
  }

  Future<Post> getPost(int postId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/posts/$postId'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 200) {
        final postData = jsonDecode(response.body);

        // Fix image URL if it's a relative path
        if (postData['image_url'] != null && postData['image_url'].toString().startsWith('/')) {
          postData['image_url'] = '$baseUrl${postData['image_url']}';
        }

        return Post.fromJson(postData);
      } else {
        throw Exception('Failed to get post: ${response.body}');
      }
    } catch (e) {
      print('Error in getPost: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<FeedResponse> getFeed({int page = 1}) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/feed?page=$page'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 200) {
        final feedData = jsonDecode(response.body);

        // Fix image URLs in posts if they're relative paths
        if (feedData['posts'] != null) {
          for (var post in feedData['posts']) {
            if (post['image_url'] != null && post['image_url'].toString().startsWith('/')) {
              post['image_url'] = '$baseUrl${post['image_url']}';
            }
          }
        }

        return FeedResponse.fromJson(feedData);
      } else {
        throw Exception('Failed to get feed: ${response.body}');
      }
    } catch (e) {
      print('Error in getFeed: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<FeedResponse> getDiscoverFeed({int page = 1}) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/discover?page=$page'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 200) {
        final feedData = jsonDecode(response.body);

        // Fix image URLs in posts if they're relative paths
        if (feedData['posts'] != null) {
          for (var post in feedData['posts']) {
            if (post['image_url'] != null && post['image_url'].toString().startsWith('/')) {
              post['image_url'] = '$baseUrl${post['image_url']}';
            }
          }
        }

        return FeedResponse.fromJson(feedData);
      } else {
        throw Exception('Failed to get discover feed: ${response.body}');
      }
    } catch (e) {
      print('Error in getDiscoverFeed: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<void> likePost(int postId) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/community/posts/$postId/like'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode != 204) {
        throw Exception('Failed to like post: ${response.body}');
      }
    } catch (e) {
      print('Error in likePost: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<void> unlikePost(int postId) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/api/community/posts/$postId/like'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode != 204) {
        throw Exception('Failed to unlike post: ${response.body}');
      }
    } catch (e) {
      print('Error in unlikePost: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<void> bookmarkPost(int postId) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/community/posts/$postId/bookmark'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode != 204) {
        throw Exception('Failed to bookmark post: ${response.body}');
      }
    } catch (e) {
      print('Error in bookmarkPost: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<void> removeBookmark(int postId) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/api/community/posts/$postId/bookmark'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode != 204) {
        throw Exception('Failed to remove bookmark: ${response.body}');
      }
    } catch (e) {
      print('Error in removeBookmark: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<FeedResponse> getBookmarkedPosts({int page = 1}) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/bookmarks?page=$page'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 200) {
        final feedData = jsonDecode(response.body);

        // Fix image URLs in posts if they're relative paths
        if (feedData['posts'] != null) {
          for (var post in feedData['posts']) {
            if (post['image_url'] != null && post['image_url'].toString().startsWith('/')) {
              post['image_url'] = '$baseUrl${post['image_url']}';
            }
          }
        }

        return FeedResponse.fromJson(feedData);
      } else {
        throw Exception('Failed to get bookmarked posts: ${response.body}');
      }
    } catch (e) {
      print('Error in getBookmarkedPosts: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<List<Post>> getUserPosts(int userId, {int page = 1}) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/users/$userId/posts?page=$page'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);

        // Fix image URLs in posts if they're relative paths
        for (var post in data) {
          if (post['image_url'] != null && post['image_url'].toString().startsWith('/')) {
            post['image_url'] = '$baseUrl${post['image_url']}';
          }
        }

        return data.map((post) => Post.fromJson(post)).toList();
      } else {
        throw Exception('Failed to get user posts: ${response.body}');
      }
    } catch (e) {
      print('Error in getUserPosts: $e');
      throw Exception('Network error: $e');
    }
  }

  // Comment endpoints
  Future<Comment> createComment(int postId, String text) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/community/comments'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'post_id': postId,
          'text': text,
        }),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 201) {
        return Comment.fromJson(jsonDecode(response.body));
      } else {
        throw Exception('Failed to create comment: ${response.body}');
      }
    } catch (e) {
      print('Error in createComment: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<List<Comment>> getPostComments(int postId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/posts/$postId/comments'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        return data.map((comment) => Comment.fromJson(comment)).toList();
      } else {
        throw Exception('Failed to get post comments: ${response.body}');
      }
    } catch (e) {
      print('Error in getPostComments: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<void> likeComment(int commentId) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/community/comments/$commentId/like'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode != 204) {
        throw Exception('Failed to like comment: ${response.body}');
      }
    } catch (e) {
      print('Error in likeComment: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<void> unlikeComment(int commentId) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/api/community/comments/$commentId/like'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode != 204) {
        throw Exception('Failed to unlike comment: ${response.body}');
      }
    } catch (e) {
      print('Error in unlikeComment: $e');
      throw Exception('Network error: $e');
    }
  }

  // Reply endpoints
  Future<Reply> createReply(int commentId, String text) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/community/replies'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'comment_id': commentId,
          'text': text,
        }),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 201) {
        return Reply.fromJson(jsonDecode(response.body));
      } else {
        throw Exception('Failed to create reply: ${response.body}');
      }
    } catch (e) {
      print('Error in createReply: $e');
      throw Exception('Network error: $e');
    }
  }

  Future<List<Reply>> getCommentReplies(int commentId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/community/comments/$commentId/replies'),
      ).timeout(const Duration(seconds: 60));

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        return data.map((reply) => Reply.fromJson(reply)).toList();
      } else {
        throw Exception('Failed to get comment replies: ${response.body}');
      }
    } catch (e) {
      print('Error in getCommentReplies: $e');
      throw Exception('Network error: $e');
    }
  }
}

File: MoneyMind_Backend/src/services/community_services.py
from sqlalchemy.orm import Session
from typing import List, Optional, Dict, Any
from sqlalchemy.sql import func
from sqlalchemy import desc, asc

from ..models.community_models import User, Post, Comment, Reply, Bookmark, Follow
from fastapi import HTTPException, UploadFile
import os
from config.settings import MEDIA_STORAGE_PATH
import uuid
import shutil

class CommunityService:
    def __init__(self, db_session: Session):
        self.db = db_session

    # User operations
    def get_user(self, user_id: int) -> Optional[User]:
        return self.db.query(User).filter(User.id == user_id).first()
    
    def get_user_by_handle(self, handle: str) -> Optional[User]:
        return self.db.query(User).filter(User.handle == handle).first()
    
    def create_user(self, username: str, handle: str, bio: str = None) -> User:
        if self.get_user_by_handle(handle):
            raise HTTPException(status_code=400, detail="Handle already exists")
        
        user = User(
            username=username,
            handle=handle,
            bio=bio
        )
        self.db.add(user)
        self.db.commit()
        self.db.refresh(user)
        return user
    
    def update_user(self, user_id: int, data: Dict[str, Any]) -> User:
        user = self.get_user(user_id)
        if not user:
            raise HTTPException(status_code=404, detail="User not found")
        
        for key, value in data.items():
            if hasattr(user, key):
                setattr(user, key, value)
        
        self.db.commit()
        self.db.refresh(user)
        return user
    
    def upload_profile_image(self, user_id: int, file: UploadFile) -> str:
        user = self.get_user(user_id)
        if not user:
            raise HTTPException(status_code=404, detail="User not found")
        
        os.makedirs(os.path.join(MEDIA_STORAGE_PATH, "profiles"), exist_ok=True)
        
        file_extension = file.filename.split(".")[-1]
        unique_filename = f"{uuid.uuid4()}.{file_extension}"
        file_path = os.path.join(MEDIA_STORAGE_PATH, "profiles", unique_filename)
        
        with open(file_path, "wb") as buffer:
            shutil.copyfileobj(file.file, buffer)
        
        user.profile_image = file_path
        self.db.commit()
        
        return file_path
    
    def follow_user(self, follower_id: int, followed_id: int) -> None:
        if follower_id == followed_id:
            raise HTTPException(status_code=400, detail="Users cannot follow themselves")
        
        existing_follow = self.db.query(Follow).filter(
            Follow.follower_id == follower_id,
            Follow.followed_id == followed_id
        ).first()
        
        if existing_follow:
            raise HTTPException(status_code=400, detail="Already following this user")
        
        follow = Follow(follower_id=follower_id, followed_id=followed_id)
        self.db.add(follow)
        self.db.commit()
    
    def unfollow_user(self, follower_id: int, followed_id: int) -> None:
        follow = self.db.query(Follow).filter(
            Follow.follower_id == follower_id,
            Follow.followed_id == followed_id
        ).first()
        
        if not follow:
            raise HTTPException(status_code=404, detail="Not following this user")
        
        self.db.delete(follow)
        self.db.commit()
    
    def get_followers(self, user_id: int) -> List[User]:
        return self.db.query(User).join(
            Follow, User.id == Follow.follower_id
        ).filter(Follow.followed_id == user_id).all()
    
    def get_following(self, user_id: int) -> List[User]:
        return self.db.query(User).join(
            Follow, User.id == Follow.followed_id
        ).filter(Follow.follower_id == user_id).all()
    
    def search_users(self, query: str, limit: int = 10) -> List[User]:
        return self.db.query(User).filter(
            (User.username.ilike(f"%{query}%")) | 
            (User.handle.ilike(f"%{query}%"))
        ).limit(limit).all()
    
    # Post operations
    def create_post(self, user_id: int, content: str, image_file: Optional[UploadFile] = None) -> Post:
        post = Post(
            user_id=user_id,
            content=content,
            has_image=bool(image_file)
        )
        
        self.db.add(post)
        self.db.flush()
        
        # Handle image upload
        if image_file:
            os.makedirs(os.path.join(MEDIA_STORAGE_PATH, "posts"), exist_ok=True)
            
            file_extension = image_file.filename.split(".")[-1]
            unique_filename = f"{uuid.uuid4()}.{file_extension}"
            file_path = os.path.join(MEDIA_STORAGE_PATH, "posts", unique_filename)
            
            with open(file_path, "wb") as buffer:
                shutil.copyfileobj(image_file.file, buffer)
            
            post.image_url = file_path
        
        self.db.commit()
        self.db.refresh(post)
        return post
    
    def get_post(self, post_id: int) -> Optional[Post]:
        return self.db.query(Post).filter(Post.id == post_id).first()
    
    def get_user_posts(self, user_id: int, page: int = 1, per_page: int = 10) -> List[Post]:
        return self.db.query(Post).filter(
            Post.user_id == user_id
        ).order_by(
            desc(Post.created_at)
        ).offset((page - 1) * per_page).limit(per_page).all()
    
    def get_feed_posts(self, user_id: int, page: int = 1, per_page: int = 10) -> List[Post]:
        followed_users = self.db.query(Follow.followed_id).filter(Follow.follower_id == user_id).all()
        followed_ids = [user[0] for user in followed_users]
        
        if user_id not in followed_ids:
            followed_ids.append(user_id)
        
        return self.db.query(Post).filter(
            Post.user_id.in_(followed_ids)
        ).order_by(
            desc(Post.created_at)
        ).offset((page - 1) * per_page).limit(per_page).all()
    
    def get_for_you_posts(self, page: int = 1, per_page: int = 10) -> List[Post]:
        return self.db.query(Post).order_by(
            desc(Post.created_at)
        ).offset((page - 1) * per_page).limit(per_page).all()
    
    def like_post(self, user_id: int, post_id: int) -> None:
        user = self.get_user(user_id)
        post = self.get_post(post_id)
        
        if not user or not post:
            raise HTTPException(status_code=404, detail="User or post not found")
        
        if user in post.liked_by:
            raise HTTPException(status_code=400, detail="Post already liked")
        
        post.liked_by.append(user)
        self.db.commit()
    
    def unlike_post(self, user_id: int, post_id: int) -> None:
        user = self.get_user(user_id)
        post = self.get_post(post_id)
        
        if not user or not post:
            raise HTTPException(status_code=404, detail="User or post not found")
        
        if user not in post.liked_by:
            raise HTTPException(status_code=400, detail="Post not liked")
        
        post.liked_by.remove(user)
        self.db.commit()
    
    def bookmark_post(self, user_id: int, post_id: int) -> None:
        # Check if bookmark exists
        existing_bookmark = self.db.query(Bookmark).filter(
            Bookmark.user_id == user_id,
            Bookmark.post_id == post_id
        ).first()
        
        if existing_bookmark:
            raise HTTPException(status_code=400, detail="Post already bookmarked")
        
        bookmark = Bookmark(user_id=user_id, post_id=post_id)
        self.db.add(bookmark)
        self.db.commit()
    
    def remove_bookmark(self, user_id: int, post_id: int) -> None:
        bookmark = self.db.query(Bookmark).filter(
            Bookmark.user_id == user_id,
            Bookmark.post_id == post_id
        ).first()
        
        if not bookmark:
            raise HTTPException(status_code=404, detail="Bookmark not found")
        
        self.db.delete(bookmark)
        self.db.commit()
    
    def get_bookmarked_posts(self, user_id: int, page: int = 1, per_page: int = 10) -> List[Post]:
        return self.db.query(Post).join(
            Bookmark, Post.id == Bookmark.post_id
        ).filter(
            Bookmark.user_id == user_id
        ).order_by(
            desc(Post.created_at)
        ).offset((page - 1) * per_page).limit(per_page).all()
    
    # Comment operations
    def create_comment(self, user_id: int, post_id: int, text: str) -> Comment:
        comment = Comment(
            user_id=user_id,
            post_id=post_id,
            text=text
        )
        self.db.add(comment)
        self.db.commit()
        self.db.refresh(comment)
        return comment
    
    def get_post_comments(self, post_id: int) -> List[Comment]:
        return self.db.query(Comment).filter(
            Comment.post_id == post_id
        ).order_by(
            Comment.created_at
        ).all()
    
    def create_reply(self, user_id: int, comment_id: int, text: str) -> Reply:
        reply = Reply(
            user_id=user_id,
            comment_id=comment_id,
            text=text
        )
        self.db.add(reply)
        self.db.commit()
        self.db.refresh(reply)
        return reply
    
    def get_comment_replies(self, comment_id: int) -> List[Reply]:
        return self.db.query(Reply).filter(
            Reply.comment_id == comment_id
        ).order_by(
            Reply.created_at
        ).all()
    
    def like_comment(self, user_id: int, comment_id: int) -> None:
        user = self.get_user(user_id)
        comment = self.db.query(Comment).filter(Comment.id == comment_id).first()
        
        if not user or not comment:
            raise HTTPException(status_code=404, detail="User or comment not found")
        
        if user in comment.liked_by:
            raise HTTPException(status_code=400, detail="Comment already liked")
        
        comment.liked_by.append(user)
        self.db.commit()
    
    def unlike_comment(self, user_id: int, comment_id: int) -> None:
        user = self.get_user(user_id)
        comment = self.db.query(Comment).filter(Comment.id == comment_id).first()
        
        if not user or not comment:
            raise HTTPException(status_code=404, detail="User or comment not found")
        
        if user not in comment.liked_by:
            raise HTTPException(status_code=400, detail="Comment not liked")
        
        comment.liked_by.remove(user)
        self.db.commit()